service.subtitles.subsceneplus/service.py

```diff
@@ -0,0 +1,304 @@
+
+class Subtitle:
+    def __init__(self, href, lang, name, no_of_files, hearing_imp, author, comment):
+        self.href = href
+
+        # first search in standard languages.
+        self.lang = lang
+        self.lang_2let = xbmc.convertLanguage(lang.strip(), xbmc.ISO_639_1)
+        self.lang_3let = xbmc.convertLanguage(lang.strip(), xbmc.ISO_639_2)
+        if self.lang_2let == "":
+            try:
+                self.lang_2let = subscene_languages[lang.strip()]['2let']
+                self.lang_3let = subscene_languages[lang.strip()]['3let']
+            except:
+                # TODO: Log the language.
+                pass
+
+        self.name = name
+        self.no_of_files = no_of_files
+        self.hearing_imp = hearing_imp
+        self.author = author
+        self.comment = comment
+        self.score = 0
+    
+    def compute_score(self, filename):
+        self.score = SequenceMatcher(None, filename, self.name).ratio()
+    
+    def __lt__(self, other):
+        return self.score > other.score
+
+    def rate(self):
+        if self.score < 0.2:
+            return "1"
+        if self.score < 0.4:
+            return "2"
+        if self.score < 0.6:
+            return "3"
+        if self.score < 0.8:
+            return "4"
+        return "5"
+
+
+def Search(item):
+    # import web_pdb; web_pdb.set_trace()
+    if item['manualsearch']:
+        allsubs = SearchMovie(item['manualsearchstring'], item['year'])
+    else:
+        allsubs = SearchMovie(item['title'], item['year'])
+
+    if allsubs is None:
+        return
+
+    
+    # Filter subtitles based on their language
+    filtered_subs = []
+    for s in allsubs:
+        subtitle = Subtitle(s[0], s[1], s[2], s[3], s[4], s[5], s[6])
+        if subtitle.lang_3let not in item['3let_language']:
+            continue
+        subtitle.compute_score(item['file_original_path'].split("/")[-1])
+        filtered_subs.append(subtitle)
+    
+    # time to remove duplicates 
+    mapped_subs = {}
+    for subtitle in filtered_subs:
+        if subtitle.href not in mapped_subs:
+            # Add subtitle to dict
+            mapped_subs[subtitle.href] = subtitle
+        else:
+            # Subtitle already exists, so we check the score, if this one has a higher score, it should replace the other one.
+            if mapped_subs[subtitle.href].score < subtitle.score:
+                mapped_subs[subtitle.href] = subtitle
+    
+    uniq_subs = []
+    for v in list(mapped_subs.values()):
+        uniq_subs.append(v)
+
+    # sort uniq_subs based on their score.
+    uniq_subs.sort()
+            
+    for subtitle in uniq_subs:
+        listitem = xbmcgui.ListItem(
+            label = subtitle.lang,              # language name for the found subtitle
+            label2 = subtitle.name,             # file name for the found subtitle
+            iconImage = subtitle.rate(),        # rating for the subtitle, string 0-5
```

Comment:
please do not use iconImage and thumbnailImage, they are both deprecated and were already removed from matrix. Use listitem.setArt({'icon': xxx, 'thumb': xxx, 'fanart': zzz, etc}